### ISSUE TITLE
Rule Request: Reduce Boolean

### ISSUE THREAD

**Issue by jpsim** [realm/SwiftLint#2583]
Now that we have `Sequence.allSatisfy` in Swift 4.2+, it's quite likely that any time we see `reduce(true)` or `reduce(false)` it could be simplified to either `.contains(...)` or `.allSatisfy(...)` to allow for bailing out early and better readability.

> 1. Why should this rule be added? Share links to existing discussion about what
>    the community thinks about this.

Should be added to improve readability and performance. See examples in https://github.com/apple/swift-evolution/blob/master/proposals/0207-containsOnly.md

> 2. Provide several examples of what _would_ and _wouldn't_ trigger violations.

```swift
// Trigger
let allNines = nums.reduce(true) { $0.0 && $0.1 == 9 } // should use allSatisfy instead
let anyNines = nums.reduce(false) { $0.0 || $0.1 == 9 } // should use contains instead
let allValid = validators.reduce(true) { $0 && $1(input) } // should use allSatisfy instead
let anyValid = validators.reduce(false) { $0 || $1(input) } // should use contains instead

// Wouldn't trigger
nums.reduce(0) { $0.0 + $0.1 }
nums.reduce(0.0) { $0.0 + $0.1 }
```

> 3. Should the rule be configurable, if so what parameters should be configurable?

We should only lint this rule when we detect a Swift version of 4.2 or later.

> 4. Should the rule be opt-in or enabled by default? Why?
>    See [README.md](../README.md#opt-in-rules) for guidelines on when to mark a rule as opt-in.

I can't think of any case where you'd want to reduce over a boolean rather than use contains or allSatisfy, so I think this should be enabled by default. I'd love to know if I'm missing some use cases though.

**Comment by marcelofabri** [comment]
This was implemented by #2675 🎉

**Comment by cjamie** [comment]
you would want to reduce boolean over a contains when you want to execute absolutely every block passed in. 

example: 


```
protocol Cond {
    var condition: Bool { get }
}

struct Hello: Cond {
    var condition: Bool {
        // some code I must execute
        return false
    }
}

struct Hello1: Cond {
    var condition: Bool {
        // some code I must execute
        return false
    }
}
struct HelloTrue: Cond {
    var condition: Bool {
        // some code I must execute
        return false
    }
}


let temp: [Cond] = [Hello(), HelloTrue(), Hello1()]
    
let result = temp.contains(where: {$0.condition})

print(result)
```

The result will exit as soon as you hit HelloTrue in the collection, and skip over any code contained within Hello1(). 


We can consider using an allSatisfy here, to loop through absolutely all elements. It would read something like 

```
let result = !temp.allSatisfy{$0.cond} 
```

or 
```
let result = temp.allSatisfy{$0.cond} == false. 
```

a reduce seems to communicate more intent, in that it will continue to go through all elements, and intuitively accumulate a final result.

temp.reduce(false) {$1.cond || $0}
